driver-core/src/main/resources/META-INF/native-image/resource-config.json

```diff
@@ -0,0 +1,7 @@
+{
+  "resources":{
+  "includes":[{
+    "pattern":"\\QMETA-INF/services/com.mongodb.spi.dns.InetAddressResolverProvider\\E"
```

Comment:
Weirdly, everything works even without this entry, but, if I collect reachability metadata, then this entry is added there.

Reply:
That's strange.  Did you find that for any other reachability metatada?  Can you suggest a hypothesis for what might be going on?

Reply:
> Did you find that for any other reachability metatada?

I haven't checked everything, nor do I understand how some of the automatically collected resources are used, but removing the shared library resources, or the `logback-test.xml` resource do have an effect: the former caused runtime errors, the latter results in log entries disappearing.

> Can you suggest a hypothesis for what might be going on?

I suspect, it's all about undocumented behavior that is intended to make things easier. I found the following:

- https://github.com/oracle/graal/issues/563#issuecomment-459121737
- https://javadoc.io/doc/org.graalvm.nativeimage/svm/20.3.10/com/oracle/svm/hosted/ServiceLoaderFeature.html

Of the two experimental options mentioned in the PR comment linked above, only `-H:-UseServiceLoaderFeature` is recognized, and it breaks `ServiceLoader` completely:

- if we don't have the `META-INF/services/com.mongodb.spi.dns.InetAddressResolverProvider` resource entry, `ServiceLoader` silently fails to locate `CustomInetAddressResolverProvider` and `CustomInetAddressResolverProvider.assertUsed()` fails 
- if we do have the `META-INF/services/com.mongodb.spi.dns.InetAddressResolverProvider`, the `ServiceLoader` fails with `java.util.ServiceConfigurationError: com.mongodb.spi.dns.InetAddressResolverProvider: Provider com.mongodb.internal.graalvm.CustomInetAddressResolverProvider not found` even if I re-collect the reachability metadata (it does not find anything new).

**P.S.** Before I found the above, I did and wrote the following:

> I had a **un**educated guess that maybe the entry is not needed because the `META-INF/services/com.mongodb.spi.dns.InetAddressResolverProvider` resource is in the same project as the one I am building, and if the resource were in a pre-built JAR that I depend on, then the resource entry would have been required.
> 
> Since you asked, I tested that guess: created a new Gradle module with the resource and `CustomInetAddressResolverProvider`, removed those from `:graalvm-native-image-app`, and added the `configuration:'archives'` dependency on the new module to `:graalvm-native-image-app`. The result is: the entry about `META-INF/services/com.mongodb.spi.dns.InetAddressResolverProvider` in `resource-config.json` is still not needed. Note that when there is no entry, the resource is actually not mentioned in the `registerResource` messages emitted by `native-image` when it builds, yet `CustomInetAddressResolverProvider` is still successfully located via `ServiceLoader`.
> 
> I don't have any other hypotheses. The behavior is really weird.